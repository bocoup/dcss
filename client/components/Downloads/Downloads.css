.dl__left_col_empty {
  border-bottom: 0px !important;
  border-top: 0px !important;
  background: white;
  min-width: 47px;
}

.dl__left_col_icon {
  border-right: 1px solid rgba(34,36,38,.1);
  border-bottom: 0px;
  border-top: 0px;
  background: white;
  min-width: 47px;
}

.dl__table tbody td {
  min-height: 45px !important;
  height: 45px !important;
}

.dl__button {
  /*background: none !important;*/
}

@media only screen and (max-width: 767px) {
  .dl__hidden-on-mobile {
    display: none !important;
  }
}
