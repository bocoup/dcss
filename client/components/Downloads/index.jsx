import React, { Component, Fragment } from 'react';
import hash from 'object-hash';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { Parser } from 'json2csv';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import {
  Button,
  Container,
  Dropdown,
  Icon,
  /*Input,*/
  Menu,
  Pagination,
  Table
} from '@components/UI';
import { getAllCohorts, getCohorts } from '@actions/cohort';
import { getHistoryForScenario } from '@actions/history';
import { getScenarios } from '@actions/scenario';
import { getUser } from '@actions/user';
import EditorMenu from '@components/EditorMenu';
import Loading from '@components/Loading';
import CSV from '@utils/csv';
import { makeHeader } from '@utils/data-table';
import shorten from '@utils/shorten';
import '../Cohorts/Cohort.css';
import '../Cohorts/DataTable.css';
import './Downloads.css';

const ROWS_PER_PAGE = 10;

function isAudioFile(input) {
  return /^audio\/\d.+\/AudioResponse/.test(input) && input.endsWith('.mp3');
}

class Downloads extends Component {
  constructor(props) {
    super(props);

    const { activePage, filter } = this.props;

    this.state = {
      isReady: false,
      activePage: activePage || 1,
      filter
    };
    this.onDownloadSearchChange = this.onDownloadSearchChange.bind(this);
    this.onPageChange = this.onPageChange.bind(this);
    this.requestDownload = this.requestDownload.bind(this);
  }

  async componentDidMount() {
    await this.props.getUser();

    if (!this.props.user.id) {
      this.props.history.push('/logout');
    } else {
      await this.props.getScenarios();

      if (this.props.user.is_super) {
        await this.props.getAllCohorts();
      } else {
        await this.props.getCohorts();
      }

      this.setState({ isReady: true });
    }
  }

  onDownloadSearchChange(options, query) {
    const re = new RegExp(_.escapeRegExp(query));
    return options.filter(option => re.test(option.text));
  }

  onPageChange(event, { activePage }) {
    this.props.history.push(`/downloads/${activePage}`);
    this.setState({
      activePage
    });
  }

  async requestDownload({ cohort, scenarioId }) {
    const { getHistoryForScenario } = this.props;
    const scenarioIds = scenarioId
      ? [scenarioId]
      : cohort.scenarios.map(v => (typeof v === 'number' ? v : v.id));
    const files = [];

    for (let id of scenarioIds) {
      const { prompts, responses } = await getHistoryForScenario(
        id,
        cohort && cohort.id
      );

      const records = responses.flat();
      records.forEach(record => {
        const {
          is_skip,
          response_id,
          response: { content },
          transcript,
          value
        } = record;
        const prompt = prompts.find(
          prompt => prompt.responseId === response_id
        );
        record.header = makeHeader(prompt, prompts);
        record.content = content || '';
        record.content += is_skip ? '(skipped)' : ` ${transcript || value}`;

        if (isAudioFile(value)) {
          record.content += ` (${location.origin}/api/media/${value})`;
        }

        if (cohort) {
          record.cohort_id = cohort.id;
        }
      });

      const fields = [
        'username',
        'header',
        'content',
        'is_skip',
        'run_id',
        'created_at',
        'ended_at',
        'type',
        'referrer_params',
        'consent_granted_by_user'
      ];

      if (cohort) {
        fields.push('cohort_id');
      }

      const file = hash({ cohort, id });
      const parser = new Parser({ fields });
      const csv = parser.parse(records);

      files.push([`${file}.csv`, csv]);
    }

    if (Object.entries(files).length === 1) {
      const [file, csv] = files[0];
      CSV.download(file, csv);
    } else {
      CSV.downloadZipAsync(files);
    }
  }

  render() {
    const { onDownloadSearchChange, onPageChange, requestDownload } = this;
    const { activePage, filter, isReady } = this.state;
    const {
      cohorts,
      /*cohortsById, */ scenarios,
      scenariosById,
      user
    } = this.props;

    if (!isReady) {
      return <Loading />;
    }

    const hasAccessToCohort = cohort => {
      if (user.is_super) {
        return true;
      }
      return cohort.users.find(({ id, roles }) => {
        return id === user.id && roles.includes('researcher');
      });
    };

    const downloadZipIcon = (
      <Icon.Group size="big" className="ig__multiple-file">
        <Icon name="file archive outline" />
        <Icon name="download" corner="top right" color="green" />
      </Icon.Group>
    );

    const downloadRunIcon = (
      <Icon.Group>
        <Icon name="file alternate outline" />
        <Icon name="download" corner="top right" color="green" />
      </Icon.Group>
    );

    const dropdownCohortSelectOptions = [
      {
        key: -1,
        value: -1,
        text: 'Find cohort'
      }
    ];

    const downloads = cohorts.reduce((accum, cohort, index) => {
      if (hasAccessToCohort(cohort)) {
        const onDownloadByCohortClick = () => {
          requestDownload({ cohort });
        };

        dropdownCohortSelectOptions.push({
          key: hash({ cohort, index }),
          value: cohort.id,
          text: shorten(cohort.name, 50)
        });

        if (filter && (filter.type !== 'cohort' || filter.id !== cohort.id)) {
          return accum;
        }

        accum.push(
          ...cohort.scenarios.map((id, index) => {
            const scenario = scenariosById[id];
            const onDownloadByScenarioRunClick = () => {
              requestDownload({ cohort, scenarioId: id });
            };

            return (
              <Table.Row
                key={hash({ ...cohort, id })}
                created_at={Date.now(cohort.created_at)}
              >
                {index === 0 ? (
                  <Table.Cell.Clickable
                    verticalAlign="top"
                    popup="Download a zip containing csv files containing responses for all scenarios in this cohort"
                    rowSpan={cohort.scenarios.length}
                    content={downloadZipIcon}
                    onClick={onDownloadByCohortClick}
                  />
                ) : null}

                <Table.Cell.Clickable
                  verticalAlign="top"
                  popup="Download a csv file containing responses to only this scenario, from this cohort"
                  content={downloadRunIcon}
                  onClick={onDownloadByScenarioRunClick}
                />
                <Table.Cell
                  verticalAlign="top"
                  className="dtr__cell-fluid-half"
                >
                  {cohort.name}
                </Table.Cell>
                <Table.Cell verticalAlign="top" className="dt__cell-truncated">
                  {scenario.title}
                </Table.Cell>
              </Table.Row>
            );
          })
        );
      }
      return accum;
    }, []);

    dropdownCohortSelectOptions.sort((a, b) => {
      return a.text < b.text;
    });

    let dropdownScenarioSelectOptions = null;
    if (user.is_super) {
      dropdownScenarioSelectOptions = [
        {
          key: -1,
          value: -1,
          text: 'Find scenario'
        }
      ];

      const scenarioRunDownloads = scenarios.reduce(
        (accum, scenario, index) => {
          const onDownloadByScenarioRunClick = () => {
            requestDownload({ scenarioId: scenario.id });
          };

          dropdownScenarioSelectOptions.push({
            key: hash({ scenario, index }),
            value: scenario.id,
            text: shorten(scenario.title, 50)
          });

          if (
            filter &&
            (filter.type !== 'scenario' || filter.id !== scenario.id)
          ) {
            return accum;
          }

          accum.push(
            <Table.Row
              key={hash({ scenario })}
              created_at={Date.now(scenario.created_at)}
            >
              <Table.Cell style={{ minWidth: '47px' }}> </Table.Cell>
              <Table.Cell.Clickable
                verticalAlign="top"
                popup="Download a csv file containing responses to only this scenario"
                content={downloadRunIcon}
                onClick={onDownloadByScenarioRunClick}
              />
              <Table.Cell
                verticalAlign="top"
                className="dtr__cell-fluid-half dt__cell-data-missing"
              ></Table.Cell>
              <Table.Cell verticalAlign="top" className="dt__cell-truncated">
                {scenario.title}
              </Table.Cell>
            </Table.Row>
          );
          return accum;
        },
        []
      );

      dropdownScenarioSelectOptions.sort((a, b) => {
        return a.text < b.text;
      });

      downloads.push(...scenarioRunDownloads);
    }

    downloads.sort((a, b) => {
      return a.props.created_at > b.props.created_at;
    });

    const downloadsPages = Math.ceil(downloads.length / ROWS_PER_PAGE);
    const downloadsIndex = (activePage - 1) * ROWS_PER_PAGE;
    const downloadsSlice = downloads.slice(
      downloadsIndex,
      downloadsIndex + ROWS_PER_PAGE
    );
    const menuItemDownloadsCount = (
      <Menu.Item key="menu-item-downloads-count">
        <Icon.Group className="em__icon-group-margin">
          {downloadRunIcon.props.children}
        </Icon.Group>
        Downloads ({downloads.length})
      </Menu.Item>
    );

    /*
    let filterDisplay = null;
    if (filter) {
      filterDisplay =
        filter.type === 'cohort'
          ? cohortsById[filter.id].name
          : scenariosById[filter.id].title;
    }
    const menuItemFilterDisplay = filter ? (
      <Menu.Item key="menu-item-filter-display">
        <Icon.Group className="em__icon-group-margin">
          {downloadRunIcon.props.children}
        </Icon.Group>
        {shorten(filterDisplay, 50)}
      </Menu.Item>
    ) : null;
  */
    const pushFilterToHistory = ({ id, type }) => {
      if (id !== -1) {
        this.props.history.push(`/downloads/${type}/${id}`);
        this.setState({ filter: { type, id } });
      }
    };

    const menuItemCohortSelect = dropdownCohortSelectOptions ? (
      <Menu.Item key="menu-item-cohort-select" style={{ width: '30%' }}>
        <Dropdown
          fluid
          item
          selection
          scrolling
          placeholder="Select cohort"
          selectOnNavigation={false}
          defaultValue={-1}
          options={dropdownCohortSelectOptions}
          search={onDownloadSearchChange}
          onChange={(_, { value: id }) => {
            pushFilterToHistory({ id, type: 'cohort' });
          }}
        />
      </Menu.Item>
    ) : null;

    const menuItemScenarioSelect = dropdownScenarioSelectOptions ? (
      <Menu.Item key="menu-item-scenario-select" style={{ width: '30%' }}>
        <Dropdown
          fluid
          item
          selection
          scrolling
          placeholder="Select scenario"
          selectOnNavigation={false}
          defaultValue={-1}
          options={dropdownScenarioSelectOptions}
          search={onDownloadSearchChange}
          onChange={(_, { value: id }) => {
            pushFilterToHistory({ id, type: 'scenario' });
          }}
        />
      </Menu.Item>
    ) : null;

    /*
    const menuItemClearSelect = filter ? (
      <Menu.Item
        key="menu-item-clear-select"
        as={NavLink}
        exact
        to="/downloads/1"
      >
        <Icon name="x" /> Clear filter
      </Menu.Item>
    ) : null;
    const menuItemDownloadsSearch = (
      <Menu.Menu key="menu-item-downloads-search" position="right">
        <Menu.Item name="Search downloads">
          <Input
            icon="search"
            placeholder="Search..."
            onChange={onDownloadSearchChange}
          />
        </Menu.Item>
      </Menu.Menu>
    );
    */

    const menuBar = (
      <EditorMenu
        type="administration"
        items={{
          left: [
            // filter ? menuItemFilterDisplay : menuItemDownloadsCount,
            menuItemDownloadsCount
          ],
          right: [
            // menuItemCohortSelect,
            // menuItemScenarioSelect,
            // menuItemClearSelect,
            // menuItemDownloadsSearch
          ]
        }}
      />
    );

    const buttonProps = { compact: true, basic: true };
    if (!filter) {
      buttonProps.disabled = true;
    }

    return (
      <Fragment>
        {menuBar}
        <Container fluid>
          <Table role="grid" unstackable striped>
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell colSpan={2}>
                  <Button
                    {...buttonProps}
                    onClick={() => {
                      this.props.history.push('/downloads');
                      this.setState({ filter: null });
                    }}
                  >
                    <Icon.Group size="large">
                      <Icon name="filter" />
                      {filter ? (
                        <Icon corner="top right" name="x" color="red" />
                      ) : null}
                    </Icon.Group>
                  </Button>
                </Table.HeaderCell>
                <Table.HeaderCell className="dtr__cell-fluid-half-th">
                  {menuItemCohortSelect.props.children}
                </Table.HeaderCell>
                <Table.HeaderCell className="dtr__cell-fluid-half-th">
                  {user.is_super
                    ? menuItemScenarioSelect.props.children
                    : 'Scenario'}
                </Table.HeaderCell>
              </Table.Row>
            </Table.Header>
            <Table.Body>
              {downloadsSlice.length ? (
                downloadsSlice
              ) : (
                <Table.Row>
                  <Table.Cell colSpan={100}>No downloads available.</Table.Cell>
                </Table.Row>
              )}
            </Table.Body>
            <Table.Footer>
              <Table.Row>
                <Table.HeaderCell colSpan="4">
                  {downloadsPages > 1 ? (
                    <Pagination
                      name="downloads"
                      siblingRange={1}
                      boundaryRange={0}
                      ellipsisItem={null}
                      firstItem={null}
                      lastItem={null}
                      activePage={activePage}
                      onPageChange={onPageChange}
                      totalPages={downloadsPages}
                    />
                  ) : null}
                </Table.HeaderCell>
              </Table.Row>
            </Table.Footer>
          </Table>
        </Container>
      </Fragment>
    );
  }
}

Downloads.propTypes = {
  activePage: PropTypes.node,
  cohorts: PropTypes.array,
  cohortsById: PropTypes.object,
  filter: PropTypes.object,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string,
    search: PropTypes.string,
    state: PropTypes.object
  }),
  match: PropTypes.shape({
    path: PropTypes.string,
    params: PropTypes.shape({
      id: PropTypes.node,
      type: PropTypes.string
    }).isRequired
  }).isRequired,
  scenarios: PropTypes.array,
  scenariosById: PropTypes.object,
  getAllCohorts: PropTypes.func,
  getCohorts: PropTypes.func,
  getScenarios: PropTypes.func,
  getHistoryForScenario: PropTypes.func,
  getUser: PropTypes.func,
  onClick: PropTypes.func,
  user: PropTypes.object
};

const mapStateToProps = (state, ownProps) => {
  const { cohorts, cohortsById, scenarios, scenariosById, user } = state;
  let { filter = null } = ownProps.location.state || {};

  if (!filter && ownProps.match.params) {
    const { id, type } = ownProps.match.params || {};

    filter = id && type ? { id: Number(id), type } : null;
  }

  let activePage = Number(ownProps.match.params.activePage) || null;

  return {
    activePage,
    cohorts,
    cohortsById,
    filter,
    scenarios,
    scenariosById,
    user
  };
};

const mapDispatchToProps = dispatch => ({
  getAllCohorts: () => dispatch(getAllCohorts()),
  getCohorts: () => dispatch(getCohorts()),
  getScenarios: () => dispatch(getScenarios()),
  getHistoryForScenario: (...params) =>
    dispatch(getHistoryForScenario(...params)),
  getUser: () => dispatch(getUser())
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(Downloads)
);
