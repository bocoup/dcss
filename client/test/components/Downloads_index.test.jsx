import {
  mounter,
  reduxer,
  snapshot,
  state
} from '../bootstrap';
import {
  unmountComponentAtNode
} from 'react-dom';
import {
  cleanup,
  fireEvent,
  render
} from '@testing-library/react';
import Downloads from '../../components/Downloads/index.jsx';

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

let container = null;
beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetch.mockImplementation(() => {
    return Promise.resolve({
      status: 200,
      json() {
        return Promise.resolve({});
      },
    });
  });
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

const sharedProps = {
  history: {
    push() {},
  },
};

test('Downloads', () => {
  expect(Downloads).toBeDefined();
});

test('Snapshot 1', () => {
  const props = {
    ...sharedProps,
  };
  const mounted = mounter(reduxer(Downloads, props, state))();
  expect(snapshot(mounted)).toMatchSnapshot();
});

/*{INJECTION}*/
