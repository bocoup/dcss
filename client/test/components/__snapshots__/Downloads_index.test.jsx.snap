// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `"<div style=\\"width: 100%; height: 90%;\\" class=\\"ui card loading__single-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder\\"><div class=\\"square image\\"></div></div></div></div>"`;

exports[`Render 1 1 2`] = `"<div style=\\"width: 100%; height: 90%;\\" class=\\"ui card loading__single-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder\\"><div class=\\"square image\\"></div></div></div></div>"`;

exports[`Render 1 1 3`] = `
<DocumentFragment>
  <div
    class="ui card loading__single-card"
    style="width: 100%; height: 90%;"
  >
    <div
      class="content loading__content"
    >
      <div
        class="ui placeholder"
      >
        <div
          class="square image"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render 1 1 4`] = `
<DocumentFragment>
  <div
    class="ui borderless icon em__height menu"
  >
    <div
      class="item"
      tabindex="0"
    >
      <i
        class="icons em__icon-group-margin"
      >
        <i
          aria-hidden="true"
          class="file alternate outline icon"
        />
        <i
          aria-hidden="true"
          class="green download top right corner icon"
        />
      </i>
      Downloads (4)
    </div>
  </div>
  <div
    class="ui fluid container"
  >
    <table
      class="ui striped unstackable table"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
            colspan="2"
          >
            <button
              class="ui basic compact disabled button"
              disabled=""
              tabindex="-1"
            >
              <i
                class="large icons"
              >
                <i
                  aria-hidden="true"
                  class="filter icon"
                />
              </i>
            </button>
          </th>
          <th
            class="dtr__cell-fluid-half-th"
          >
            <div
              aria-expanded="false"
              class="ui fluid item search selection scrolling dropdown"
              role="combobox"
            >
              <input
                aria-autocomplete="list"
                autocomplete="off"
                class="search"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                aria-atomic="true"
                aria-live="polite"
                class="text"
                role="alert"
              >
                Find cohort
              </div>
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                class="menu transition"
                role="listbox"
              >
                <div
                  aria-checked="true"
                  aria-selected="true"
                  class="active selected item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Find cohort
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    A New Cohort That Exists Within Inline Props
                  </span>
                </div>
              </div>
            </div>
          </th>
          <th
            class="dtr__cell-fluid-half-th"
          >
            <div
              aria-expanded="false"
              class="ui fluid item search selection scrolling dropdown"
              role="combobox"
            >
              <input
                aria-autocomplete="list"
                autocomplete="off"
                class="search"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                aria-atomic="true"
                aria-live="polite"
                class="text"
                role="alert"
              >
                Find scenario
              </div>
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                class="menu transition"
                role="listbox"
              >
                <div
                  aria-checked="true"
                  aria-selected="true"
                  class="active selected item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Find scenario
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Multiplayer Scenario
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Some Other Scenario
                  </span>
                </div>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
          created_at="1577836800050"
        >
          <td
            aria-label="Download a zip containing csv files containing responses for all scenarios in \\"A New Cohort That Exists Within Inline Props\\""
            class="top aligned"
            rowspan="2"
            style="cursor: pointer;"
          >
            <i
              class="big icons ig__multiple-file"
            >
              <i
                aria-hidden="true"
                class="file archive outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            aria-label="Download a csv file containing responses to only \\"Multiplayer Scenario\\", from the cohort \\"A New Cohort That Exists Within Inline Props\\""
            class="top aligned"
            style="cursor: pointer;"
          >
            <i
              class="icons"
            >
              <i
                aria-hidden="true"
                class="file alternate outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            class="top aligned dtr__cell-fluid-half"
          >
            A New Cohort That Exists Within Inline Props
          </td>
          <td
            class="top aligned dt__cell-truncated"
          >
            Multiplayer Scenario
          </td>
        </tr>
        <tr
          class=""
          created_at="1577836800050"
        >
          <td
            aria-label="Download a csv file containing responses to only \\"Some Other Scenario\\", from the cohort \\"A New Cohort That Exists Within Inline Props\\""
            class="top aligned"
            style="cursor: pointer;"
          >
            <i
              class="icons"
            >
              <i
                aria-hidden="true"
                class="file alternate outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            class="top aligned dtr__cell-fluid-half"
          >
            A New Cohort That Exists Within Inline Props
          </td>
          <td
            class="top aligned dt__cell-truncated"
          >
            Some Other Scenario
          </td>
        </tr>
        <tr
          class=""
          created_at="1577836800050"
        >
          <td
            class=""
            style="min-width: 47px;"
          >
             
          </td>
          <td
            aria-label="Download a csv file containing responses to only \\"Multiplayer Scenario\\""
            class="top aligned"
            style="cursor: pointer;"
          >
            <i
              class="icons"
            >
              <i
                aria-hidden="true"
                class="file alternate outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            class="top aligned dtr__cell-fluid-half dt__cell-data-missing"
          />
          <td
            class="top aligned dt__cell-truncated"
          >
            Multiplayer Scenario
          </td>
        </tr>
        <tr
          class=""
          created_at="1577836800050"
        >
          <td
            class=""
            style="min-width: 47px;"
          >
             
          </td>
          <td
            aria-label="Download a csv file containing responses to only \\"Some Other Scenario\\""
            class="top aligned"
            style="cursor: pointer;"
          >
            <i
              class="icons"
            >
              <i
                aria-hidden="true"
                class="file alternate outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            class="top aligned dtr__cell-fluid-half dt__cell-data-missing"
          />
          <td
            class="top aligned dt__cell-truncated"
          >
            Some Other Scenario
          </td>
        </tr>
      </tbody>
      <tfoot
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
            colspan="4"
          />
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;

exports[`Render 2 1 1`] = `"<div style=\\"width: 100%; height: 90%;\\" class=\\"ui card loading__single-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder\\"><div class=\\"square image\\"></div></div></div></div>"`;

exports[`Render 2 1 2`] = `"<div style=\\"width: 100%; height: 90%;\\" class=\\"ui card loading__single-card\\"><div class=\\"content loading__content\\"><div class=\\"ui placeholder\\"><div class=\\"square image\\"></div></div></div></div>"`;

exports[`Render 2 1 3`] = `
<DocumentFragment>
  <div
    class="ui card loading__single-card"
    style="width: 100%; height: 90%;"
  >
    <div
      class="content loading__content"
    >
      <div
        class="ui placeholder"
      >
        <div
          class="square image"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render 2 1 4`] = `
<DocumentFragment>
  <div
    class="ui borderless icon em__height menu"
  >
    <div
      class="item"
      tabindex="0"
    >
      <i
        class="icons em__icon-group-margin"
      >
        <i
          aria-hidden="true"
          class="file alternate outline icon"
        />
        <i
          aria-hidden="true"
          class="green download top right corner icon"
        />
      </i>
      Downloads (2)
    </div>
  </div>
  <div
    class="ui fluid container"
  >
    <table
      class="ui striped unstackable table"
      role="grid"
    >
      <thead
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
            colspan="2"
          >
            <button
              class="ui basic compact disabled button"
              disabled=""
              tabindex="-1"
            >
              <i
                class="large icons"
              >
                <i
                  aria-hidden="true"
                  class="filter icon"
                />
              </i>
            </button>
          </th>
          <th
            class="dtr__cell-fluid-half-th"
          >
            <div
              aria-expanded="false"
              class="ui fluid item search selection scrolling dropdown"
              role="combobox"
            >
              <input
                aria-autocomplete="list"
                autocomplete="off"
                class="search"
                tabindex="0"
                type="text"
                value=""
              />
              <div
                aria-atomic="true"
                aria-live="polite"
                class="text"
                role="alert"
              >
                Find cohort
              </div>
              <i
                aria-hidden="true"
                class="dropdown icon"
              />
              <div
                class="menu transition"
                role="listbox"
              >
                <div
                  aria-checked="true"
                  aria-selected="true"
                  class="active selected item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    Find cohort
                  </span>
                </div>
                <div
                  aria-checked="false"
                  aria-selected="false"
                  class="item"
                  role="option"
                  style="pointer-events: all;"
                >
                  <span
                    class="text"
                  >
                    A New Cohort That Exists Within Inline Props
                  </span>
                </div>
              </div>
            </div>
          </th>
          <th
            class="dtr__cell-fluid-half-th"
          >
            Scenario
          </th>
        </tr>
      </thead>
      <tbody
        class=""
      >
        <tr
          class=""
          created_at="1577836800150"
        >
          <td
            aria-label="Download a zip containing csv files containing responses for all scenarios in \\"A New Cohort That Exists Within Inline Props\\""
            class="top aligned"
            rowspan="2"
            style="cursor: pointer;"
          >
            <i
              class="big icons ig__multiple-file"
            >
              <i
                aria-hidden="true"
                class="file archive outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            aria-label="Download a csv file containing responses to only \\"Multiplayer Scenario\\", from the cohort \\"A New Cohort That Exists Within Inline Props\\""
            class="top aligned"
            style="cursor: pointer;"
          >
            <i
              class="icons"
            >
              <i
                aria-hidden="true"
                class="file alternate outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            class="top aligned dtr__cell-fluid-half"
          >
            A New Cohort That Exists Within Inline Props
          </td>
          <td
            class="top aligned dt__cell-truncated"
          >
            Multiplayer Scenario
          </td>
        </tr>
        <tr
          class=""
          created_at="1577836800150"
        >
          <td
            aria-label="Download a csv file containing responses to only \\"Some Other Scenario\\", from the cohort \\"A New Cohort That Exists Within Inline Props\\""
            class="top aligned"
            style="cursor: pointer;"
          >
            <i
              class="icons"
            >
              <i
                aria-hidden="true"
                class="file alternate outline icon"
              />
              <i
                aria-hidden="true"
                class="green download top right corner icon"
              />
            </i>
          </td>
          <td
            class="top aligned dtr__cell-fluid-half"
          >
            A New Cohort That Exists Within Inline Props
          </td>
          <td
            class="top aligned dt__cell-truncated"
          >
            Some Other Scenario
          </td>
        </tr>
      </tbody>
      <tfoot
        class=""
      >
        <tr
          class=""
        >
          <th
            class=""
            colspan="4"
          />
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;
